<a href="https://colab.research.google.com/github/holps-7/NLP-weekly/blob/master/Magnum_Opus.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import nltk
nltk.download('all')

[nltk_data] Downloading collection 'all'
[nltk_data]    | 
[nltk_data]    | Downloading package abc to /root/nltk_data...
[nltk_data]    |   Unzipping corpora/abc.zip.
[nltk_data]    | Downloading package alpino to /root/nltk_data...
[nltk_data]    |   Unzipping corpora/alpino.zip.
[nltk_data]    | Downloading package biocreative_ppi to
[nltk_data]    |     /root/nltk_data...
[nltk_data]    |   Unzipping corpora/biocreative_ppi.zip.
[nltk_data]    | Downloading package brown to /root/nltk_data...
[nltk_data]    |   Unzipping corpora/brown.zip.
[nltk_data]    | Downloading package brown_tei to /root/nltk_data...
[nltk_data]    |   Unzipping corpora/brown_tei.zip.
[nltk_data]    | Downloading package cess_cat to /root/nltk_data...
[nltk_data]    |   Unzipping corpora/cess_cat.zip.
[nltk_data]    | Downloading package cess_esp to /root/nltk_data...
[nltk_data]    |   Unzipping corpora/cess_esp.zip.
[nltk_data]    | Downloading package chat80 to /root/nltk_data...
[nltk_data]    |   Unzipp

True

In [2]:
#Alchemy
doc1='''The Great Work (Latin: Magnum opus) is an alchemical term for the process 
of working with the prima materia to create the philosopher's stone. 
It has been used to describe personal and spiritual transmutation in the Hermetic tradition, 
attached to laboratory processes and chemical color changes, used as a model for the individuation process, 
and as a device in art and literature. The magnum opus has been carried forward in New Age and 
neo-Hermetic movements which sometimes attached new symbolism and significance to the processes.'''

In [3]:
#Hermeticism
doc2='''The term Great Work (magnum opus) is a term used in Hermeticism and 
occult traditions descended from it, most prominently Thelema. 
The Great Work signifies the spiritual path towards self-transcendence in its entirety. 
This is the process of bringing unconscious complexes into the conscious awareness, 
in order to integrate them back into oneself. Accomplishing the Great Work, 
symbolized as the creation of the Philosopher's Stone, represents the culmination of the 
spiritual path, the attainment of enlightenment, or the rescue of the 
human soul from the unconscious forces which bind it.'''

In [4]:
from nltk.stem import LancasterStemmer
LS=LancasterStemmer()

In [5]:
Stem1=[LS.stem(i) for i in doc1.split()]
Stem1=" ".join(Stem1)
Stem1

"the gre work (latin: magn opus) is an alchem term for the process of work with the prim mater to cre the philosopher's stone. it has been us to describ person and spirit transmut in the hermet tradition, attach to lab process and chem col changes, us as a model for the individu process, and as a dev in art and literature. the magn op has been carry forward in new ag and neo-hermetic mov which sometim attach new symbol and sign to the processes."

In [6]:
Stem2=[LS.stem(i) for i in doc2.split()]
Stem2=" ".join(Stem2)
Stem2

"the term gre work (magnum opus) is a term us in hermet and occult tradit descend from it, most promin thelema. the gre work sign the spirit path toward self-transcendence in it entirety. thi is the process of bring unconscy complex into the conscy awareness, in ord to integr them back into oneself. accompl the gre work, symbol as the cre of the philosopher's stone, repres the culmin of the spirit path, the attain of enlightenment, or the rescu of the hum soul from the unconscy forc which bind it."

In [7]:
from sklearn.feature_extraction.text import CountVectorizer
vect=CountVectorizer(binary=True)

In [8]:
corpus = [Stem1,Stem2]

In [9]:
vect.fit(corpus)

CountVectorizer(analyzer='word', binary=True, decode_error='strict',
                dtype=<class 'numpy.int64'>, encoding='utf-8', input='content',
                lowercase=True, max_df=1.0, max_features=None, min_df=1,
                ngram_range=(1, 1), preprocessor=None, stop_words=None,
                strip_accents=None, token_pattern='(?u)\\b\\w\\w+\\b',
                tokenizer=None, vocabulary=None)

In [10]:
vocab=vect.vocabulary_
len(vocab)

92

In [11]:
from sklearn.metrics.pairwise import cosine_similarity
similarity = cosine_similarity(vect.transform([Stem1]).toarray(), vect.transform([Stem2]).toarray())
similarity

array([[0.38602432]])

In [12]:
from sklearn.feature_extraction.text import CountVectorizer
vect=CountVectorizer(binary=True)

In [13]:
corpus = [doc1,doc2]

In [14]:
vect.fit(corpus)

CountVectorizer(analyzer='word', binary=True, decode_error='strict',
                dtype=<class 'numpy.int64'>, encoding='utf-8', input='content',
                lowercase=True, max_df=1.0, max_features=None, min_df=1,
                ngram_range=(1, 1), preprocessor=None, stop_words=None,
                strip_accents=None, token_pattern='(?u)\\b\\w\\w+\\b',
                tokenizer=None, vocabulary=None)

In [15]:
vocab=vect.vocabulary_
len(vocab)

95

In [16]:
from sklearn.metrics.pairwise import cosine_similarity
similarity = cosine_similarity(vect.transform([doc1]).toarray(), vect.transform([doc2]).toarray())
similarity

array([[0.33353871]])

# Results:
## With Stemming: 
  * Vocabulary Size = 92
  * Cosine Similarity = 0.38602432

## Without Stemming: 
  * Vocabulary Size = 95
  * Cosine Similarity = 0.33353871